src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -1,30 +1,164 @@
-                    GetSystemTimeWithLeapSecondsHandling(&time);
-                    return CreateDateTimeFromSystemTime(in time);
+                    return new DateTime((osTicks + FileTimeOffset) | KindUtc);
                 }
 
-                return new DateTime(((ulong)(GetSystemTimeAsFileTime() + FileTimeOffset)) | KindUtc);
+                // If our cache has been populated and we're within its validity period, rely
+                // solely on the cache. We use unsigned arithmetic below so that if the system
+                // clock is set backward, the resulting value will integer underflow and cause
+                // a cache miss, forcing us down the slow path.
+
+                if (s_leapSecondCache is LeapSecondCache cache)
+                {
+                    ulong deltaTicksFromStartOfMonth = osTicks - cache.OSFileTimeTicksAtStartOfMonth;
+                    if (deltaTicksFromStartOfMonth < cache.CacheValidityPeriodInTicks)
+                    {
+                        return new DateTime(deltaTicksFromStartOfMonth + cache.DotNetDateDataAtStartOfMonth); // includes UTC marker
+                    }
+                }
+
+                // Move uncommon fallback pack into its own subroutine so that register
+                // allocation in get_UtcNow can stay as optimized as possible.
+
+                return Fallback(osTicks);
+                static DateTime Fallback(ulong osTicks)
+                {
+                    // If we reached this point, one of the following is true:
+                    //   a) the cache hasn't yet been initialized; or
+                    //   b) the month has changed since the last call to UtcNow; or
+                    //   c) the current time is 23:59:59 or 23:59:60 on the last day of the month.
+                    //
+                    // In cases (a) and (b), we'll update the cache. In case (c), we
+                    // pessimistically assume we might be inside a leap second, so we
+                    // won't update the cache.
+
+                    DateTime utcNow = FromFileTimeLeapSecondsAware((long)osTicks);
+                    DateTime oneSecondBeforeMonthEnd = new DateTime(utcNow.Year, utcNow.Month, DaysInMonth(utcNow.Year, utcNow.Month), 23, 59, 59, DateTimeKind.Utc);
+
+                    if (utcNow < oneSecondBeforeMonthEnd)
+                    {
+                        // It's not yet 23:59:59 on the last day of the month, so update the cache.
+                        // It's ok for multiple threads to do this concurrently as long as the write
+                        // to the static field is published *after* the cache's instance fields have
+                        // been populated.
+
+                        DateTime startOfMonth = new DateTime(utcNow.Year, utcNow.Month, 1, 0, 0, 0, DateTimeKind.Utc);
+                        Debug.Assert(startOfMonth <= utcNow);
+
+                        Volatile.Write(ref s_leapSecondCache, new LeapSecondCache
```

Comment:
Under CLR memory model reference writes to static fields always have release semantics so I think `Volatile.Write` is not needed.

Reply:
Writes to static fields do not always have release semantics on some of the platforms that Mono runs on. I think it is a good idea to have Volatile.Writes for clarity, even when it is not strictly required.

Reply:
> I think it is a good idea to have Volatile.Writes for clarity, even when it is not strictly required.

I agree - I put it there for clarity. An alternative design would be to mark the static field itself _volatile_, then we wouldn't need the call to `Volatile.Write`. But since correctness is predicated on only the write (not the read) being volatile, I thought it would be cleaner to make that fact explicit. If consensus is simply to mark the static field itself _volatile_ I can do that as well.

Reply:
I think we have hundreds of `s_foo = new ...` assignments in this repo outside of class constructors.  I do not see why this assignment is special in any way and inconsistency does not help clarity.

Reply:
> I think we have hundreds of s_foo = new ... assignments in this repo outside of class constructors.

Interestingly enough, I ran a quick Regex to find these, and the code below was one of the first matches. Whoops. 😳

https://github.com/dotnet/runtime/blob/162b33e19eb5a3434867541be691e1082039b130/src/libraries/System.Net.Http/src/System/Net/Http/Headers/EntityTagHeaderValue.cs#L26-L38

Reply:
> I do not see why this assignment is special in any way and inconsistency does not help clarity.

We try to write new code "correctly" and fix up existing code opportunistically.

Reply:
> Whoops.

Yeah, volatile or not, that's a bug.

Reply:
"Correctly" is a vague term until the alternative memory model is properly defined, documented, and developers are aware of it.

Reply:
ECMA

Reply:
JIT has never implemented the ECMA memory model (e.g., regarding volatile reads), bug reports (mine included) have been ignored for years, developers have been told that JIT actually implemented something different, so it is hard to take this argument seriously.

Reply:
The _coreclr_ JIT has implemented a stronger memory model.  I'm not sure why that invalidates coding to the published open standard.

Reply:
Wrong. It is stronger in some aspects, but weaker in other aspects, e.g., dropping volatile reads or treating volatile reads as non-volatile ones.  That is why "coding to the published open standard" does not guarantee correctness and proves my point:

>"Correctly" is a vague term until the alternative memory model is properly defined, documented, and developers are aware of it.

Reply:
Can you please link to the GitHub issues citing places it's weaker?

Reply:
Searching for "volatile read" finds, for example, #6280 and #10619.  One of the bugs I filed in 2008 is [here](https://devdiv.visualstudio.com/defaultcollection/DevDiv/_workitems/edit/90798/).  Let me quote the official resolution:

>Currently [the ECMA spec requirement] is not the MS implementation [...] We should update the ECMA spec such that the MS behavior is explicitly allowed [...] Then this bug can be resolved as "By Design".

If we are serious about following the ECMA spec, can we prioritize fixing long-standing bugs or finally updating the model?

Reply:
> for example, #6280 and #10619

Both of which remain open and tagged as bugs.  Believe it or not, production quality software still has bugs.

> One of the bugs I filed in 2008 is here

It sounds like the bug there is in the specification itself.  So that bug should be fixed.

Reply:
Incorrect handling of volatile reads for 20 years is not just some corner case — it is obvious contract violation.  For that reason alone we should prioritize properly defining the contract (i.e. correctness rules) over making the code look "correct".

Reply:
Please open whatever issues you feel need to be opened.  None of that contradicts being explicit about where volatile writes should happen, even if the _coreclr JIT_ happens to already do so.  I'm going to stop responding to this comment chain.  Thanks.